{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56a14e86",
   "metadata": {},
   "source": [
    "An open reading frame (ORF) is one which starts from the start codon and ends by stop codon, without any other stop codons in between. Thus, a candidate protein string is derived by translating an open reading frame into amino acids until a stop codon is reached.\n",
    "Reading frames: 3 per direction (start from 1st, 2nd, 2rd base), also check reverse complement!\n",
    "\n",
    "Given: A DNA string s\n",
    " of length at most 1 kbp in FASTA format.\n",
    "\n",
    "Return: Every distinct candidate protein string that can be translated from ORFs of s\n",
    ". Strings can be returned in any order.\n",
    "\n",
    "Tasks: \n",
    "\n",
    "-read in FASTA\n",
    "-get reverse complement\n",
    "-extract triplets\n",
    "-given ORF & starting pos -->test if peptide exists\n",
    "    -->return list of peptide codons\n",
    "-translate list of codons to protein\n",
    "-function for getting all valid proteins from 1 ORF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d345bf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0af224fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find open reading frames\n",
    "\n",
    "# loop through strand, find start codons\n",
    "# work in slices of three from each base\n",
    "# if slice == start codon\n",
    "    #check if there is a stop codon between start and end of string (substring from start codon [current_base:], find function?, if result is within substring) \n",
    "    #find--> returns index, so can be my if conditional (does find return something?); use elif to check for all stop codons\n",
    "    #create a substring and add to to orf list?\n",
    "\n",
    "    --> do for each reading frame; (do for revc and strand if going base by base; or go in triplets and three reading frames in both directions)\n",
    "\n",
    "\n",
    "    #alternative would be: translate for each triplet, then only keep proteins that start with m and end with stop\n",
    "\n",
    "def open_reading_frames(rna_strand)\n",
    "'''for an RNA strand, finds all possible open reading frames and adds them to a list as substrings\n",
    "\n",
    "an open reading frame is a codon sequence between a start and a stop codon\n",
    "in every direction there are 3 reading frames, so 6 in total for RNA and reverse complement'''\n",
    "    orf_list = () #creating an empty orf list\n",
    "\n",
    "    for base in rna_strand: #looping through all bases\n",
    "\n",
    "        if rna_strand [base:(base + 3)] == \"ATG\" #check if the current codon (base and next two) is a start codon\n",
    "            potential_orf = rna_strand [base:] #create a substring that goes from start codon until the end of strand\n",
    "\n",
    "            if potential_orf.find(\"TAG\") > -1: #if potential_orf contains a stop codon\n",
    "                #if there is no other stop codon before\n",
    "                #if potential_orf.find(\"TGA\") > \n",
    "                # what to do if no other stop codon, then index will be -1\n",
    "                stop_index = potential_orf.find(\"TAG\") #save the starting index of the stop codon\n",
    "                orf = rna_strand[base:(stop_index + 3)] #slice rna_strand from start to stop codon\n",
    "                orf_list.append(orf) #adding orf to orf list\n",
    "\n",
    "            elif potential_orf.find(\"TGA\") > -1:\n",
    "                stop_index = potential_orf.find(\"TGA\") #save the starting index of the stop codon\n",
    "                orf = rna_strand[base:(stop_index + 3)] #slice rna_strand from start to stop codon\n",
    "                orf_list.append(orf) #adding orf to orf list\n",
    "\n",
    "            elif potential_orf.find(\"TAA\") > -1:\n",
    "                stop_index = potential_orf.find(\"TAA\") #save the starting index of the stop codon\n",
    "                orf = rna_strand[base:(stop_index + 3)] #slice rna_strand from start to stop codon\n",
    "                orf_list.append(orf) #adding orf to orf list\n",
    "\n",
    "\n",
    "            # now do the same for revc\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56df0d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check what find will return if the sequence is not contained\n",
    "\n",
    "string = \"abcd\"\n",
    "not_substring = \"e\"\n",
    "\n",
    "string.find(not_substring)\n",
    "\n",
    "# still pending, the internet says find will return -1 if the substring is not contained"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4cf15d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if find will only return first or all instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91e96e46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate orfs into proteins\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "carpentries",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
